language/ast.py

```diff
@@ -0,0 +1,677 @@
+    description = Field(type=Description)
+    body = Field(type=BaseStateBody)
+
+
+class SessionDecl(Declaration):
+
+    parameters = Field(type=FormalDecl.list)
+    identifier = Field(type=UnqualifiedID)
+    aspects = Field(type=SessionAspects)
+    declarations = Field(type=LocalDecl.list)
+    states = Field(type=State.list)
+    end_identifier = Field(type=UnqualifiedID)
+
+
+class Package(RFLXNode):
+
+    identifier = Field(type=UnqualifiedID)
+    declarations = Field(type=Declaration.list)
+    end_identifier = Field(type=UnqualifiedID)
+
+
+@abstract
+class IntegerTypeDef(TypeDef):
+    """
+    Base class for all integer type definitions
+    """
+
+
+class Aspect(RFLXNode):
+
+    identifier = Field(type=UnqualifiedID)
+    value = Field(type=Expr)
+
+
+class RangeTypeDef(IntegerTypeDef):
+
+    lower = Field(type=Expr)
+    upper = Field(type=Expr)
+    size = Field(type=Aspect)
+
+
+class ModularTypeDef(IntegerTypeDef):
+
+    mod = Field(type=Expr)
+
+
+@abstract
+class AbstractMessageTypeDef(TypeDef):
+    """
+    Base class for message type definitions
+    """
+
+
+class NullMessageTypeDef(AbstractMessageTypeDef):
+    pass
+
+
+class TypeDerivationDef(TypeDef):
+
+    base = Field(type=ID)
+
+
+class ArrayTypeDef(TypeDef):
+
+    element_type = Field(type=ID)
+
+
+@abstract
+class EnumerationDef(TypeDef):
+    """
+    Base class for enumeration definitions
+    """
+
+
+class PositionalEnumerationDef(EnumerationDef):
+
+    elements = Field(type=UnqualifiedID.list)
+
+
+class ElementValueAssoc(RFLXNode):
+    """
+    Element/value association
+    """
+
+    identifier = Field(type=UnqualifiedID)
+    literal = Field(type=NumericLiteral)
+
+
+class NamedEnumerationDef(EnumerationDef):
+
+    elements = Field(type=ElementValueAssoc.list)
+
+
+class EnumerationTypeDef(TypeDef):
+
+    elements = Field(type=EnumerationDef)
+    aspects = Field(type=Aspect.list)
+
+
+class Then(RFLXNode):
+    """
+    Link to field
+    """
+
+    target = Field(type=AbstractID)
+    aspects = Field(type=Aspect.list)
+    condition = Field(type=Expr)
+
+
+class NullComponent(RFLXNode):
+
+    then = Field(type=Then)
+
+
+class Component(RFLXNode):
+    """
+    Message component
+    """
+
+    identifier = Field(type=UnqualifiedID)
+    type_identifier = Field(type=ID)
+    aspects = Field(type=Aspect.list)
+    condition = Field(type=Expr)
+    thens = Field(type=Then.list)
+
+
+class Components(RFLXNode):
+    """
+    Message components
+    """
+
+    initial_component = Field(type=NullComponent)
+    components = Field(type=Component.list)
+
+
+@abstract
+class BaseChecksumVal(RFLXNode):
+    """
+    Base class for checksum values
+    """
+
+
+class ChecksumVal(BaseChecksumVal):
+    """
+    Single checksum value
+    """
+
+    data = Field(type=Expr)
+
+
+class ChecksumValueRange(BaseChecksumVal):
+    """
+    Checksum value range
+    """
+
+    lower = Field(type=Expr)
+    upper = Field(type=Expr)
```

Comment:
Same as for ranged types.

Reply:
Changed in 1546458cc4c71b2f4f4c8c0fa843eea8c81e0e77